File: gesttravaux-symfony/src/Entity/Categorie.php
<?php

namespace App\Entity;

use App\Repository\CategorieRepository;
use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;
use Doctrine\ORM\Mapping as ORM;

/**
 * Categorie entity - Work categories
 *
 * Represents different types of construction/renovation work categories
 * (e.g., Plumbing, Electrical, Masonry, etc.)
 *
 * @author GestTravaux Pro
 */
#[ORM\Entity(repositoryClass: CategorieRepository::class)]
#[ORM\Table(name: 'categories')]
class Categorie
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column(name: 'idCategorie')]
    private ?int $id = null;

    #[ORM\Column(length: 45)]
    private ?string $type = null;

    /**
     * @var Collection<int, Prestation>
     */
    #[ORM\OneToMany(targetEntity: Prestation::class, mappedBy: 'categorie')]
    private Collection $prestations;

    /**
     * @var Collection<int, Entrepreneur>
     */
    #[ORM\ManyToMany(targetEntity: Entrepreneur::class, mappedBy: 'categories')]
    private Collection $entrepreneurs;

    public function __construct()
    {
        $this->prestations = new ArrayCollection();
        $this->entrepreneurs = new ArrayCollection();
    }

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getType(): ?string
    {
        return $this->type;
    }

    public function setType(string $type): static
    {
        $this->type = $type;

        return $this;
    }

    /**
     * @return Collection<int, Prestation>
     */
    public function getPrestations(): Collection
    {
        return $this->prestations;
    }

    public function addPrestation(Prestation $prestation): static
    {
        if (!$this->prestations->contains($prestation)) {
            $this->prestations->add($prestation);
            $prestation->setCategorie($this);
        }

        return $this;
    }

    public function removePrestation(Prestation $prestation): static
    {
        if ($this->prestations->removeElement($prestation)) {
            // set the owning side to null (unless already changed)
            if ($prestation->getCategorie() === $this) {
                $prestation->setCategorie(null);
            }
        }

        return $this;
    }

    /**
     * @return Collection<int, Entrepreneur>
     */
    public function getEntrepreneurs(): Collection
    {
        return $this->entrepreneurs;
    }

    public function addEntrepreneur(Entrepreneur $entrepreneur): static
    {
        if (!$this->entrepreneurs->contains($entrepreneur)) {
            $this->entrepreneurs->add($entrepreneur);
            $entrepreneur->addCategory($this);
        }

        return $this;
    }

    public function removeEntrepreneur(Entrepreneur $entrepreneur): static
    {
        if ($this->entrepreneurs->removeElement($entrepreneur)) {
            $entrepreneur->removeCategory($this);
        }

        return $this;
    }

    public function __toString(): string
    {
        return $this->type ?? '';
    }
}

File: gesttravaux-java/src/main/java/com/immosync/gesttravaux/models/Categorie.java
package com.immosync.gesttravaux.models;

import jakarta.persistence.*;
import java.util.HashSet;
import java.util.Set;

/**
 * Entity representing a category of work/services.
 * Categories can be associated with multiple entrepreneurs and prestations.
 *
 * @author GestTravaux Pro
 * @version 1.0
 */
@Entity
@Table(name = "CATEGORIES")
public class Categorie {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "idCategorie")
    private Integer idCategorie;

    @Column(name = "type", length = 45, nullable = false)
    private String type;

    @ManyToMany(mappedBy = "categories", fetch = FetchType.LAZY)
    private Set<Entrepreneur> entrepreneurs = new HashSet<>();

    @OneToMany(mappedBy = "categorie", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private Set<Prestation> prestations = new HashSet<>();

    /**
     * Default constructor required by JPA.
     */
    public Categorie() {
    }

    /**
     * Constructor with type.
     *
     * @param type the category type
     */
    public Categorie(String type) {
        this.type = type;
    }

    // Getters and Setters

    /**
     * Gets the category ID.
     *
     * @return the category ID
     */
    public Integer getIdCategorie() {
        return idCategorie;
    }

    /**
     * Sets the category ID.
     *
     * @param idCategorie the category ID
     */
    public void setIdCategorie(Integer idCategorie) {
        this.idCategorie = idCategorie;
    }

    /**
     * Gets the category type.
     *
     * @return the category type
     */
    public String getType() {
        return type;
    }

    /**
     * Sets the category type.
     *
     * @param type the category type
     */
    public void setType(String type) {
        this.type = type;
    }

    /**
     * Gets the entrepreneurs associated with this category.
     *
     * @return set of entrepreneurs
     */
    public Set<Entrepreneur> getEntrepreneurs() {
        return entrepreneurs;
    }

    /**
     * Sets the entrepreneurs associated with this category.
     *
     * @param entrepreneurs set of entrepreneurs
     */
    public void setEntrepreneurs(Set<Entrepreneur> entrepreneurs) {
        this.entrepreneurs = entrepreneurs;
    }

    /**
     * Gets the prestations associated with this category.
     *
     * @return set of prestations
     */
    public Set<Prestation> getPrestations() {
        return prestations;
    }

    /**
     * Sets the prestations associated with this category.
     *
     * @param prestations set of prestations
     */
    public void setPrestations(Set<Prestation> prestations) {
        this.prestations = prestations;
    }

    /**
     * Adds a prestation to this category.
     *
     * @param prestation the prestation to add
     */
    public void addPrestation(Prestation prestation) {
        prestations.add(prestation);
        prestation.setCategorie(this);
    }

    /**
     * Removes a prestation from this category.
     *
     * @param prestation the prestation to remove
     */
    public void removePrestation(Prestation prestation) {
        prestations.remove(prestation);
        prestation.setCategorie(null);
    }

    @Override
    public String toString() {
        return type;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Categorie)) return false;
        Categorie categorie = (Categorie) o;
        return idCategorie != null && idCategorie.equals(categorie.getIdCategorie());
    }

    @Override
    public int hashCode() {
        return getClass().hashCode();
    }
}
